### PR TITLE
feat: check for conflicting files

### PR DESCRIPTION
**Pull request by AlexanderHOtt**
# Check for conflicting files

- [x] I reviewed linter warnings + errors, resolved formatting, types and other issues related to my work
- [x] The PR title follows the convention we established [conventional-commit](https://www.conventionalcommits.org/en/v1.0.0/)
- [-] I performed a functional test on my final commit

It works fine if given a project name, but I couldn't figure out how to test it with `.` as the name because `pnpm start .` fails with
 ```
> next start "."

'next' is not recognized as an internal or external command,
operable program or batch file.
```
when inside a test folder.

---

_[Short summary/description of story/feature/bugfix]_
Checks if existing files will be overwritten and adds the option to overwrite files if there are conflicts.
```
ℹ project is not empty, but no conflicting files found, continuing...
```
If there are conflicts
```
? Warning: project already exists and has conflicting files:

  project\.env-example
  project\.eslintrc.json
  project\next-env.d.ts
  project\next.config.mjs
  project\package.json
  project\public\favicon.ico
  project\README.md
  project\env\client-env.mjs
  project\env\env-schema.mjs
  project\env\server-env.mjs
  project\pages\index.tsx
  project\pages\_app.tsx
  project\styles\globals.css
  project\tsconfig.json
  project\.gitignore

 How would you like to proceed? (Use arrow keys)
❯ Clear directory
  Abort installation
  Overwrite files (dangerous)
```
---

## Screenshots

_[Screenshots]_
image
image

💯

### CODE REVIEW

**Comment by AlexanderHOtt** [comment]
Fix for #230

File: src/helpers/scaffoldProject.ts [hunk 1]
```diff
@@ -32,24 +32,35 @@ export const scaffoldProject = async ({
...
+        choices: [
+          { name: "Clear directory", value: "clear", short: "Clear" },
+          { name: "Abort installation", value: "abort", short: "Abort" },
+          {
+            name: "Overwrite files (dangerous)",
+            value: "overwrite",
+            short: "Overwrite",
+          },
+        ],
+        default: "clear",
```
Comment (juliusmarminge) [inline src/helpers/scaffoldProject.ts:70]: the default should be `abort` so that people don't accidentally press enter and get all their files removed.

```suggestion
        default: "abort",
```

File: src/helpers/scaffoldProject.ts [hunk 2]
```diff
@@ -26,30 +27,62 @@ export const scaffoldProject = async ({
...
-          name: "overwriteDir",
-          type: "confirm",
-          message: `${chalk.redBright.bold("Warning:")} ${chalk.cyan.bold(
-            projectName,
-          )} already exists and isn't empty. Do you want to overwrite it?`,
-          default: false,
-        },
-      );
-      if (!overwriteDir) {
+      const { overwriteDir } = await inquirer.prompt<{ overwriteDir: string }>({
```
Comment (juliusmarminge) [inline src/helpers/scaffoldProject.ts:55]: Making this string type safe will help preventing typos if modified later.

```suggestion
      const { overwriteDir } = await inquirer.prompt<{ overwriteDir: "clear" | "abort" | "overwrite" }>({
```

File: src/utils/getAllFiles.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ * Find all files inside a dir, recursively.
+ * https://gist.github.com/kethinov/6658166?permalink_comment_id=3379782#gistcomment-3379782
+ * https://gist.github.com/kethinov/6658166?permalink_comment_id=2389484#gistcomment-2389484
+ * @function getAllFiles
+ * @param {string} dir Directory path string.
+ * @param {string} base Base path string.
+ * @return {string[]} Array with all the relative paths that are inside the directory.
+ */
```
Comment (juliusmarminge) [inline src/utils/getAllFiles.ts:12]: Not a fan of jsdoc and we don't use it anywhere else. Using typescript is clear enough and inlined. Leave the reference link if it is taken externally.

**Review by juliusmarminge — COMMENTED** [review]
Not available to test this today so will have another look over the weekend/next week. 

I can see some edge-cases on the top of my head. E.g. user is in a git repo which is not caught but this logic. We should probably extend the initializeGit to handle this. @JacobMGEvans sent me this a while back that may be helpful: [cloudflare/wrangler2 - git-client.ts](https://github.com/cloudflare/wrangler2/blob/main/packages/wrangler/src/git-client.ts)

**Comment by AlexanderHOtt** [comment]
> user is in a git repo which is not caught but this logic
The `git init` should fail but that won't overwrite any files.

File: src/helpers/scaffoldProject.ts [hunk 4]
```diff
@@ -26,30 +27,64 @@ export const scaffoldProject = async ({
   const spinner = ora(`Scaffolding in: ${projectDir}...\n`).start();
 
   if (fs.existsSync(projectDir)) {
-    if (fs.readdirSync(projectDir).length === 0) {
+    const userFilesSet = new Set(getAllFiles(projectDir));
+    const srcFiles = [...getAllFiles(srcDir), ".gitignore"]; // template files
```
Comment (juliusmarminge) [inline src/helpers/scaffoldProject.ts:31]: this means we are just checking against the base app without any addons. If there is files that would conflict with any of these other ones these would not be matched and potentially removed by accident

File: src/helpers/scaffoldProject.ts [hunk 5]
```diff
@@ -26,30 +27,64 @@ export const scaffoldProject = async ({
...
+    const conflictingFiles = srcFiles.filter((element) =>
```
Comment (juliusmarminge) [inline src/helpers/scaffoldProject.ts:32]: userFilesSet might have e.g. `prisma/schema.prisma` which would not be in srcFiles and therefore not in conflictingFiles. Continuing to scaffold and letting the user pick prisma would override the existing file which is not what we want.

**Review by juliusmarminge — CHANGES_REQUESTED** [review]
Sorry for the delay on this review...

I believe there are a few gotchas which have to be addressed before we can merge this feature

**Comment by juliusmarminge** [comment]
Closed as inactive
